borsh-derive-internal/src/attribute_helpers.rs

```diff
@@ -18,3 +23,120 @@ pub fn contains_initialize_with(attrs: &[Attribute]) -> Option<Path> {
+                                "`use_discriminant` is the only supported attribute for `borsh`",
+                            )
+                            .to_compile_error(),
+                        ));
+                    }
+
+                    if let syn::Data::Struct(ref _data) = derive_input.data {
+                        return Err(TokenStream::from(
+                            syn::Error::new(
+                                derive_input.ident.span(),
+                                "borsh (use_discriminant=<bool>) does not support structs",
+                            )
+                            .to_compile_error(),
+                        ));
+                    }
+
+                    let value = value.to_token_stream().to_string();
+                    return match value.as_str() {
+                        "true" => Ok(Some(true)),
+                        "false" => Ok(Some(false)),
+                        _ => {
+                            return Err(TokenStream::from(
+                                syn::Error::new(
+                                    derive_input.ident.span(),
+                                    "`use_discriminant` accept only `true` or `false`",
+                                )
+                                .to_compile_error(),
+                            ));
+                        }
+                    };
+                }
+            }
+        }
+    }
+    Ok(None)
+}
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use quote::quote;
+    #[test]
+    fn test_check_use_discriminant() {
+        let item_enum: DeriveInput = syn::parse2(quote! {
+            #[derive(BorshDeserialize, Debug)]
+            #[borsh(use_discriminant = false)]
+            enum AWithUseDiscriminantFalse {
+                X,
+                Y,
+            }
+        })
+        .unwrap();
+        let actual = check_use_discriminant(item_enum).unwrap();
+
+        assert_eq!(actual, Some(false));
+    }
+
+    #[test]
+    fn test_check_use_discriminant_true() {
+        let item_enum: DeriveInput = syn::parse2(quote! {
+            #[derive(BorshDeserialize, Debug)]
+            #[borsh(use_discriminant = true)]
+            enum AWithUseDiscriminantFalse {
+                X,
+                Y,
+            }
+        })
+        .unwrap();
+        let actual = check_use_discriminant(item_enum).unwrap();
+
+        assert_eq!(actual, Some(true));
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_check_use_discriminant_wrong_value() {
+        let item_enum: DeriveInput = syn::parse2(quote! {
+            #[derive(BorshDeserialize, Debug)]
+            #[borsh(use_discriminant = 111)]
+            enum AWithUseDiscriminantFalse {
+                X,
+                Y,
+            }
+        })
+        .unwrap();
+        let actual = check_use_discriminant(item_enum);
+        assert!(actual.is_err());
+    }
+    #[test]
+    #[should_panic]
```

Comment:
`#[should_panic]` here was invalidly used for test.
the panic was not due to expected business logic error, but due to struct definition syntax error:
```rust
 struct AWithUseDiscriminantFalse {
                X,
                Y,
            }
 ```
 => 
 ```rust

            struct AWithUseDiscriminantFalse {
                x: X,
                y: Y,
            }

 ```
 Also, add assert on actual error text:
 ```rust 
         insta::assert_snapshot!(actual.unwrap_err().to_token_stream().to_string());

 ```